File: src/004/p004.scala
object Main {
  def isPalindrome(num: Int): Boolean = {
    num.toString == num.toString.reverse
  }

  def solve(): Int = {
    val maximum = (
      for {
        x <- (100 to 999)
        y <- (100 to 999)
      } yield(x * y)
    ).filter(num => isPalindrome(num)).max

    maximum
  }

  def main(args: Array[String]): Unit = {
    val result = solve()
    println(result)
  }
}

File: src/004/p004.cpp
#include <iostream>
using namespace std;

int reverse(int n) {
    int v = n, rv = 0;
    do {
        rv = (rv*10) + (v%10);
        v = v/10;
    } while (v > 0);
    return rv;
}

int solve() {
    int v = 0;
    for (int i = 999; i > 99; i--) {
        for (int j = 999; j > i-100; j--) {
            v = i * j;
            if (v == reverse(v)) {
                return v;
            }
        }
    }
    return 0;
}

int main() {
    int result = solve();
    cout << result << endl;
}
